crates/vizia_core/src/systems/accessibility.rs

```diff
@@ -0,0 +1,224 @@
+                                - editor.buffer().metrics().font_size as f64,
+                            x1: bounds.x as f64 + line.line_w as f64,
+                            y1: bounds.y as f64 + line.line_y as f64
+                                - editor.buffer().metrics().font_size as f64
+                                + line_height,
+                        });
+                        line_node.text_direction = if line.rtl {
+                            Some(TextDirection::RightToLeft)
+                        } else {
+                            Some(TextDirection::LeftToRight)
+                        };
+
+                        let mut character_lengths = Vec::with_capacity(line.glyphs.len());
+                        let mut character_positions = Vec::with_capacity(line.glyphs.len());
+                        let mut character_widths = Vec::with_capacity(line.glyphs.len());
+
+                        // Get the actual text in the line
+                        let first_glyph_pos =
+                            line.glyphs.first().map(|glyph| glyph.start).unwrap_or_default();
+                        let last_glyph_pos =
+                            line.glyphs.last().map(|glyph| glyph.end).unwrap_or_default();
+
+                        let mut line_text = text[first_glyph_pos..last_glyph_pos].to_owned();
+
+                        let word_lengths = line_text
+                            .unicode_words()
+                            .map(|word| word.len() as u8)
+                            .collect::<Vec<_>>();
+
+                        let mut line_length = 0;
+
+                        for glyph in line.glyphs.iter() {
+                            let length = (glyph.end - glyph.start) as u8;
+
+                            line_length += length as usize;
+
+                            let position = glyph.x;
+                            let width = glyph.w;
+
+                            character_lengths.push(length);
+                            character_positions.push(position);
+                            character_widths.push(width);
+                        }
+
+                        // Cosmic strips the newlines but accesskit needs them so we append them back in if line ended originally ended with a newline
+                        // If the last glyph position is equal to the end of the buffer line then this layout run is the last one and ends in a newline.
+                        if last_glyph_pos == line.text.len() {
+                            line_text += "\n";
+                            character_lengths.push(1);
+                            character_positions.push(line.line_w);
+                            character_widths.push(0.0);
+                        }
+
+                        // FIXME: The screen reader gets out of sync with the cursor due to affinity
+
+                        // TODO: Might need to append any spaces that were stripped during layout. This can be done by
```

Comment:
I'll have to double check but it seemed to me that lines which wrap can sometimes have the trailing space removed when you query the text in the line, which is done by taking the slice of text from the hard wrapped line (what cosmic gives you in a run) from the start glyph to the end glyph of that run (because the glyphs take in to account soft wrap). Most of the pain of this stems from how cosmic keeps track of lines of text. Accesskit needs a node per line, whether it's soft or hard wrapped.